### ISSUE TITLE
Revisit conditions under which Async002 is fired

### ISSUE THREAD

**Issue by jamesqo** [dotnet/roslyn-analyzers#1484]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

AsyncMethodNamesShouldEndInAsync

#### Description

The old `RenameAsyncAnalyzer` in `Unfactored` fires under the following conditions:

- Return type is `Task` or `Task<T>`
- Method name ends with `Async`
- Method is not `override` or `virtual`

In implementing the IOperation-based analyzer, I propose changing these to the following criteria:

- [ ] Return type is any task-like type (e.g. `ValueTask<T>`)
- [ ] Method name ends with `Async`, using ordinal string comparison (by default, `EndsWith(string)` will do a CurrentCulture comparison)
- [ ] Method is not `override` (why shouldn't a virtual method be renamed?)
  - When offering a codefix for a virtual/abstract method, we should also offer to rename all overrides of that method present in the solution.

**Comment by jamesqo** [comment]
Related: We should also consider detecting when the method name is slightly off because of a typo, e.g. `ConnectAsycn`, and offer `ConnectAsync` instead of `ConnectAsycnAsync` as a code fix. @333fred has Roslyn implemented any typo correction algorithms of this sort before?

**Comment by 333fred** [comment]
I don't believe we've implemented anything like this before. Some suggestions:

1. Offer to rename if the original definition is in this assembly.
2. Public names are a bit dicey here. This could get really noisy if the user has shipped an assembly and can't change the names of their methods, but at the same time this analyzer is most useful on public methods before they're shipped. I think we might need a better story for baselining an existing solution before turning on (ie, only analyze newly added methods).
3. We definitely need to ignore test methods for this.
4. We definitely need to invoke the IDE rename feature for this, so that all instances are taken care of. This rename could also potentially introduce overload resolution issues, so we'll definitely need to think about how to best handle that.

**Comment by jamesqo** [comment]
@333fred 

> I think we might need a better story for baselining an existing solution before turning on (ie, only analyze newly added methods).

That is a valid point. However, I think if the library maintainer already has a bunch of methods not suffixed with `Async` then to keep things consistent (s)he may not want to suffix future methods with `Async`. I think it would be best for us to liberally apply the fix, and for maintainers of such projects to just disable the analyzer project-wide.

> We definitely need to ignore test methods for this.

Definitely, that's a huge thing that I overlooked. How do you propose I detect test methods, though? Off the top of my head I can think of a couple ways to do this in a "clean" fashion:

- Check if current assembly is referencing assemblies that house test method attributes for popular test frameworks (e.g. `xunit.core`). If so, don't fire the analyzer at all within that assembly.
- Check if the async method has any attributes. If so, don't fire the analyzer for that method. (will lead to some, but hopefully small number of false negatives)
- Combination of the above two (don't fire when the assemblies are referenced and at least one attribute is present)

Which do you think is best?

> We definitely need to invoke the IDE rename feature for this, so that all instances are taken care of. This rename could also potentially introduce overload resolution issues, so we'll definitely need to think about how to best handle that.

Do you mean `Renamer.RenameSymbolAsync`? That 's what I used in my fixer. As for overload resolution / name conflicts, opened https://github.com/dotnet/roslyn-analyzers/issues/1485

**Comment by frankshearar** [comment]
For tests - the user can just opt to not use the analyzer for a test project.

On a different note - having just implemented something like this analyzer for a team in Microsoft (I'm frsheara if you're an FTE) - you should also require an Async suffix for methods that return awaitable types in general. For example,

````csharp
[AsyncMethodBuilder(typeof(AsyncTaskMethodBuilder))]
public class MyTask<T> : Task<T>
{
    public MyTask(T value)
        : base(() => value)
    {
    }
}

public async MyTask<int> ThisShouldHaveaSuffix => new MyTask<int>(42);
````

**Comment by frankshearar** [comment]
Aaaand I just found out that Microsoft.VisualStudio.Threading.Analyzers does all that, and more.

**Comment by sharwell** [comment]
We should just remove this rule, since it's superseded by [VSTHRD200](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD200.md).

**Comment by mavasani** [comment]
I think I have deleted all Async00XX rules from this repo.
